File: src/main/java/net/cjsah/console/command/context/CommandContextBuilder.java
package net.cjsah.console.command.context;

import net.cjsah.console.command.Command;
import net.cjsah.console.command.Dispatcher;
import net.cjsah.console.command.source.CommandSource;
import net.cjsah.console.command.tree.CommandNode;

import java.util.LinkedHashMap;
import java.util.Map;

public class CommandContextBuilder {
    private final Map<String, ParsedNodeResult<?>> arguments;
    private final CommandSource<?> source;
    private final Dispatcher dispatcher;
    private final CommandNode rootNode;
    private Command command = null;
    private IntRange range;

    public CommandContextBuilder(final Dispatcher dispatcher, final CommandSource<?> source, final CommandNode rootNode, final int start) {
        this.dispatcher = dispatcher;
        this.source = source;
        this.rootNode = rootNode;
        this.arguments = new LinkedHashMap<>();
        this.range = new IntRange(start);
    }

    public CommandContextBuilder withArgument(String name, ParsedNodeResult<?> argument) {
        this.arguments.put(name, argument);
        return this;
    }

    public CommandContextBuilder withCommand(Command command) {
        this.command = command;
        return this;
    }

    public CommandContextBuilder withRange(IntRange range) {
        this.range = new IntRange(this.range, range);
        return this;
    }

    public CommandContextBuilder copy() {
        CommandContextBuilder builder = new CommandContextBuilder(this.dispatcher, this.source, this.rootNode, this.range.getStart());
        builder.command = this.command;
        builder.arguments.putAll(this.arguments);
        builder.range = this.range;
        return builder;
    }

    public CommandContext build() {
        return new CommandContext(this.arguments, this.source, this.command);
    }

    public CommandSource<?> getSource() {
        return this.source;
    }

    public IntRange getRange() {
        return this.range;
    }

}

File: src/main/kotlin/net/cjsah/bot/console/command/context/CommandContextBuilder.kt
package net.cjsah.bot.console.command.context

import net.cjsah.bot.console.command.Command
import net.cjsah.bot.console.command.CommandSource
import net.cjsah.bot.console.command.Dispatcher
import net.cjsah.bot.console.command.tree.CommandNode
import kotlin.math.max
import kotlin.math.min

class CommandContextBuilder(
    private val dispatcher: Dispatcher,
    private var source: CommandSource,
    private val rootNode: CommandNode,
    start: Int
) {
    private val arguments = LinkedHashMap<String, ParsedNodeResult<*>>()
    private var command: Command? = null
    private var range: IntRange = IntRange(start, start)

    fun withArgument(name: String, argument: ParsedNodeResult<*>): CommandContextBuilder {
        arguments[name] = argument
        return this
    }

    fun withCommand(command: Command?): CommandContextBuilder {
        this.command = command
        return this
    }

    fun withRange(range: IntRange): CommandContextBuilder {
        this.range = encompassing(this.range, range)
        return this
    }

    fun copy(): CommandContextBuilder {
        val copy = CommandContextBuilder(dispatcher, source, rootNode, range.first)
        copy.command = command
        copy.arguments.putAll(arguments)
        copy.range = range
        return copy
    }

    fun build(): CommandContext {
        return CommandContext(arguments, command)
    }

    fun getSource() = source

    fun getRange() = range

    private fun encompassing(a: IntRange, b: IntRange) = IntRange(min(a.first, b.first), max(a.last, b.last))

}